archive/2025-11-08-code-execution-with-mcp-building-more-efficient-ai-agents-anthropic.md

```diff
@@ -0,0 +1,313 @@
+
+## **Code execution with MCP improves context efficiency**
+
+With code execution environments becoming more common for agents, a solution is to present MCP servers as code APIs rather than direct tool calls. The agent can then write code to interact with MCP servers. This approach addresses both challenges: agents can load only the tools they need and process data in the execution environment before passing results back to the model.
+
+There are a number of ways to do this. One approach is to generate a file tree of all available tools from connected MCP servers. Here's an implementation using TypeScript:
+
+```
+servers
+├── google-drive
+│   ├── getDocument.ts
+│   ├── ... (other tools)
+│   └── index.ts
+├── salesforce
+│   ├── updateRecord.ts
+│   ├── ... (other tools)
+│   └── index.ts
+└── ... (other servers)
+```
+
+Copy
+
+Then each tool corresponds to a file, something like:
+
+```
+// ./servers/google-drive/getDocument.ts
+import { callMCPTool } from "../../../client.js";
+
+interface GetDocumentInput {
+  documentId: string;
+}
+
+interface GetDocumentResponse {
+  content: string;
+}
+
+/* Read a document from Google Drive */
+export async function getDocument(input: GetDocumentInput): Promise<GetDocumentResponse> {
+  return callMCPTool<GetDocumentResponse>('google_drive__get_document', input);
+}
+```
+
+Copy
+
+Our Google Drive to Salesforce example above becomes the code:
+
+```
+// Read transcript from Google Docs and add to Salesforce prospect
+import * as gdrive from './servers/google-drive';
+import * as salesforce from './servers/salesforce';
+
+const transcript = (await gdrive.getDocument({ documentId: 'abc123' })).content;
+await salesforce.updateRecord({
+  objectType: 'SalesMeeting',
+  recordId: '00Q5f000001abcXYZ',
+  data: { Notes: transcript }
+});
+```
+
+Copy
+
+The agent discovers tools by exploring the filesystem: listing the `./servers/`directory to find available servers (like `google-drive` and `salesforce`), then reading the specific tool files it needs (like `getDocument.ts` and `updateRecord.ts`) to understand each tool's interface. This lets the agent load only the definitions it needs for the current task. This reduces the token usage from 150,000 tokens to 2,000 tokens—a time and cost saving of 98.7% **.**
+
+Cloudflare [published similar findings](https://blog.cloudflare.com/code-mode/), referring to code execution with MCP as “Code Mode." The core insight is the same: LLMs are adept at writing code and developers should take advantage of this strength to build agents that interact with MCP servers more efficiently.
+
+## **Benefits of code execution with MCP**
+
+Code execution with MCP enables agents to use context more efficiently by loading tools on demand, filtering data before it reaches the model, and executing complex logic in a single step. There are also security and state management benefits to using this approach.
```

Comment:
This is great. BUT: we are losing the cached checkpoint with all tools loaded up.